### ISSUE TITLE
Automatic volume breaking

### ISSUE THREAD

**Issue by josteinaj** [nlbdev/pipeline#20]
_From @bertfrees on November 17, 2016 12:54_

Wish/claim from NLB:

- All volumes are about the same size
- The first text page in a volume is always full (volume does not begin with the end of a chapter)
- Do not split a &lt;p&gt; in two volumes, move it to the new volume.

_Copied from original issue: nlbdev/pipeline-mod-nlb#58_

**Comment by josteinaj** [comment]
_From @bertfrees on November 23, 2016 13:34_

A part of it is already supported, and just needs to be tested. Another part (notably not splitting a &lt;p&gt; in two volumes) is not supported yet.

See https://github.com/daisy/pipeline-mod-braille/issues/136.

**Comment by josteinaj** [comment]
A suggestion for a volume breaking algorithm:
1. If at page 70, there is a headline, break just before that headline
2. If at page 70, there is no headline in the two preceding or two following pages, then break after the last full paragraph on that page
3. Otherwise do steps 1-2 for page 69
4. Otherwise do steps 1-2 for page 68
5. ...
6. Otherwise do steps 1-2 for page 4
7. Otherwise break after last full paragraph on page 70
8. Otherwise break after last full text node on page 70
9. Otherwise break after page 70

**Comment by josteinaj** [comment]
*Moved from Trello board:*

Volume breaking: currently breaks arbitrarily. Make rules for division on `<h>` or `<p>` according to better defined rules. For instance, if the last `<h>` is such and such far away, divide on the next `<p>`.

**Comment by josteinaj** [comment]
*from Trello-board:*

For reference regarding braille CSS rules: http://braillespecs.github.io/braille-css/20161201/#volume-breaking-rules

**Comment by bertfrees** [comment]
From Mats:

> - If a book consist of more than the set max page number, it must be split into volumes
>     * the max page number is a limitation, not a goal
> - The volume splitting shall seek to reflect the structure of the book, by ensuring that central structural "highlights" are used as splitting points:       In books with parts and chapters,
>     - parts will be top level splitting points (except the first part, which may start after front matter in first volume). If the content between this top level splitting points are > max page number, further splitting is necessary:
>     - chapters are second level split points (except first chapter in each part): If two or more chapters are < max page number, they may be put in the same volume (never start a new chapter inside a volume, unless all of the chapter fits inside the volume).
>     - if chapter > max page number the chapter must be split in two or more volumes:
>         - seeking a balanced page number between these volumes
>         - preferably split at headings on a lower level,
>         - but never inside a paragraph (if possible)
>         - and never after a odd page / right page (unless if split at lower level headings)
>         
>       In books with chapters,
>     - chapters are top level splitting points:
>     - If two or more chapters are < max page number, they may share volume (seeking a balanced volume page number throughout the book)  
>     - Never start a new chapter inside a volume, unless all the content of the chapter fits inside the volume
>     - if chapter > max page number, the chapter must be split in two or more volumes,
>         - seeking a balanced page number between these volumes
>         - preferably split at headings on a lower level,
>         - but never inside a paragraph (if possible)    
>         - and never after a odd page / right page (unless if split at lower level headings)

**Comment by josteinaj** [comment]
Related issues:
- ~~#63~~
- ~~#68~~
- ~~#80~~

**Comment by josteinaj** [comment]
I think we can close this issue now. After all of @bertfrees' fixes, volume breaking seems to work pretty well now. We'll still have to fine-tune how it works using CSS and some pre-processing, but we're quite happy with how volume breaking works now.
